pallets/pool-system/src/tranches.rs

```diff
@@ -1394,12 +1577,2336 @@ where
+				},
+			};
+
+			let replace_res = tranches.replace(1, input, SECS_PER_YEAR);
+			assert_eq!(
+				replace_res,
+				Err(DispatchError::Other(
+					"Invalid following tranche type. This should be catched somewhere else."
+				))
+			);
+		}
+
+		#[test]
+		fn validate_insert_works() {
+			let mut tranches = default_tranches();
+
+			let tranche_id: TrancheId = [
+				103u8, 57, 22, 242, 127, 45, 18, 102, 173, 154, 105, 163, 156, 150, 75, 194,
+			];
+			let min_risk_buffer = Perquintill::from_rational(4u64, 5);
+			let int_per_sec = Rate::saturating_from_integer(2)
+				/ Rate::saturating_from_integer(SECS_PER_YEAR)
+				+ One::one();
+
+			// verify returns valid when interest greater than tranch following new tranche
+			let new_tranche = tranches
+				.create_tranche(
+					3,
+					tranche_id,
+					TrancheType::NonResidual {
+						interest_rate_per_sec: int_per_sec,
+						min_risk_buffer: min_risk_buffer,
+					},
+					Some(5u32),
+					SECS_PER_YEAR,
+				)
+				.unwrap();
+
+			assert!(tranches.validate_insert(1, &new_tranche).is_ok());
+			// verify error when tranche new_tranche is following  has lower interest
+			assert!(tranches.validate_insert(1, &new_tranche).is_ok());
+
+			// verify error returned when interest less than tranch following new tranche
+			let int_per_sec = Rate::saturating_from_rational(1, 100)
+				/ Rate::saturating_from_integer(SECS_PER_YEAR)
+				+ One::one();
+			let new_tranche = tranches
+				.create_tranche(
+					3,
+					tranche_id,
+					TrancheType::NonResidual {
+						interest_rate_per_sec: int_per_sec,
+						min_risk_buffer: min_risk_buffer,
+					},
+					Some(5u32),
+					SECS_PER_YEAR,
+				)
+				.unwrap();
+
+			assert!(tranches.validate_insert(1, &new_tranche).is_err());
+
+			// verify ok when tranche new_tranche is following has higher interest rate
+			assert!(tranches.validate_insert(2, &new_tranche).is_ok())
+		}
+
+		#[test]
+		fn remove_tranches_works() {
+			let mut tranches = default_tranches();
+			let tranches_pre_removal = default_tranches();
+
+			// attempt to remove outside of bounds
+			assert!(tranches.remove(3).is_err());
+
+			// remove middle tranche
+			assert_ok!(tranches.remove(1));
+			assert_eq!(tranches.num_tranches(), 2);
+			assert!(tranches.get_tranche(TrancheLoc::Index(2)).is_none());
+			assert_eq!(
+				tranches_pre_removal
+					.get_tranche(TrancheLoc::Index(2))
+					.unwrap(),
+				tranches.get_tranche(TrancheLoc::Index(1)).unwrap()
+			);
+
+			// remove residual tranche
+			assert_ok!(tranches.remove(0));
+			assert_eq!(tranches.num_tranches(), 1);
+			assert!(tranches.get_tranche(TrancheLoc::Index(1)).is_none());
+			assert_eq!(
+				tranches_pre_removal
+					.get_tranche(TrancheLoc::Index(2))
+					.unwrap(),
+				tranches.get_tranche(TrancheLoc::Index(0)).unwrap()
+			);
+
+			// remove last remaining tranche
+			assert_ok!(tranches.remove(0));
+			assert!(tranches.num_tranches().is_zero());
+			assert!(tranches.remove(0).is_err());
```

Comment:
Good catch! Fixed in 31d91c62f5c2c6bd25795e4608df59053055d714

Reply:
> While looking through this I realized it should NOT be possible to remove a tranche if
> 
> * balance is != 0
> * tranche is residual tranche

Note: Needs to be applied to replacing as well.

Reply:
> > While looking through this I realized it should NOT be possible to remove a tranche if
> > 
> > * balance is != 0
> > * tranche is residual tranche
> 
> Note: Needs to be applied to replacing as well.

Answering myself: This is actually also covered by https://github.com/centrifuge/centrifuge-chain/commit/31d91c62f5c2c6bd25795e4608df59053055d714 since `replace(at)` calls `remove(at)` ensuring the invariance.